### ISSUE TITLE
yield vs. yield call

### ISSUE THREAD

**Issue by anthonychung14** [redux-saga/redux-saga#905]
being able to compose sagas is a powerful paradigm, but I find myself running into stylistic questions when breaking out sub-tasks.

oftentimes these are saga tasks that contains logic and boils down to 'do we make a server call or not'

in some cases, this involves a separate set of server calls, store-access via select, etc and I want the parent saga to block

TLDR: is there a functional difference between the following code block?

```
function* test() {
    while (true) {
        yield take('ACTION')
        const result = yield call(shouldUpdate);
        // const result = yield shouldUpdate();

        if (result.condition) //do thing
    }
}

function* shouldUpdate() {
     try {
        const value = yield call(API) 
        return value
     }
} catch
```

**Comment by jimbol** [comment]
`yield shouldUpdate();` bypasses the declarative goodies that redux sagas give you.  For example, when I'm unit testing, I'll often test the result of the yield.

```
expect(effect.next().value).to.deep.equal(call(shouldUpdate));
```

As opposed to testing that the output of `shouldUpdate`.

```
expect(effect.next().value).to.deep.equal(shouldUpdate());
```

In the second example, we have to stub or run `shouldUpdate` which might involve a bunch of set up and a whole separate effect run.

I'm sure there are other reasons.

**Comment by anthonychung14** [comment]
sweet thank you @jimbol

**Comment by jimbol** [comment]
You're welcome!

**Comment by bristoljon** [comment]
I'm gonna post my question here as it's the most relevant thread I've found for it:

I have an `apiRequest` saga that eventually yields a server response or throws an error and I need to use it inside another saga like so:

```
try {
  const response = yield call(apiRequest, { meta: { endpoint } })
  console.log('response', response)
}
catch (error) {
 console.log('error', error)
}
```
However when I do this, `apiRequest` never returns i.e. neither console log gets hit and it just hangs.

If I try `yield* call(apiRequest, { meta: { endpoint } })`, I get `response undefined`

If I call `apiRequest` directly like: `yield* apiRequest({ meta: { endpoint } })` (which is not ideal for testing), I get the desired result of the server response (or an error on server error).

I understand that `yield*` passes control to the yielded iterator until no more values are returned but shouldn't `yield* call()` do the same?

Simplified `apiRequest` in case my problem is here (although as mentioned this code works as expected when called directly).. Essentially this is a wrapper around `apiFetch` which returns a promise

```
function* apiRequest({ meta, payload = {} }) {
    const { endpoint, params } = meta
    const config = endpoints[endpoint]
    const url = formatUrl(config.url, params)
    yield put(requestFetching(endpoint, params))
    let response
    let error
    try {
      response = yield call(apiFetch, url, config.method, payload)
      yield put(requestFetched(endpoint, params, response))
    }
    catch (e) {
      error = e
      yield put(requestFailed(endpoint, params, e))
    }
    if (!error) return response
    throw error
  }
```

TL;DR: What is the difference between `yield* <saga>` and `yield* call(<saga>)` (And if you're feeling generous, `yield call(<saga>`)

**Comment by Andarist** [comment]
> I understand that yield* passes control to the yielded iterator until no more values are returned but shouldn't yield* call() do the same?

No, because result of the `call` is not an iterator, but an object (called effect). `yield*` cannot pass control to an object.

> TL;DR: What is the difference between yield* <saga> and yield* call(<saga>) (And if you're feeling generous, yield call(<saga>)

Ive already covered difference between "yield* <saga> and yield* call(<saga>", so ill focus on the latter. "yield call(<saga>)" yields result of the `call` to the caller (an object which is kinda an effect description). When using `redux-saga` its the caller, because its the one which executes generators given to it. So when saga receives an effect (by calling `iterator.next()` under the hood) it checks its type and executes it for u. Thats why in tests you can just iterate over ur sagas and receive those objects (and compare them to the expected effects), but NO action, no API call etc is made. Its because you iterate outside of the redux-saga's interpreter.

>   const response = yield call(apiRequest, { meta: { endpoint } })

This should certainly work. Could u share ur problem on something like codesandbox.io? Ill try to help you, but I need a working/broken code to work with :)

**SIDENOTE**
it would be better to structure this like that for readability:
```diff
function* apiRequest({ meta, payload = {} }) {
    const { endpoint, params } = meta
    const config = endpoints[endpoint]
    const url = formatUrl(config.url, params)
    yield put(requestFetching(endpoint, params))
-    let response
-    let error
    try {
-      response = yield call(apiFetch, url, config.method, payload)
+      const response = yield call(apiFetch, url, config.method, payload)
      yield put(requestFetched(endpoint, params, response))
+      return response
    }
    catch (e) {
-      error = e
      yield put(requestFailed(endpoint, params, e))
+      throw e
    }
-    if (!error) return response
-    throw error
}
```

**Comment by bristoljon** [comment]
Hi, thanks for your quick reply (and the sidenote - not sure why I structured it like that now!)

I see what you mean now about `yield* call()` and that explains the `object is not a generator` type error I got at one point.

I will try and replicate the problem on sandbox and let you know when it's done. Cheers

**Comment by dagadbm** [comment]
and he was never seen again

**Comment by bristoljon** [comment]
@dagadbm @Andarist  I'm back :-) Sorry never got round to providing that sandbox and certainly haven't come across the issue since so probably safe to say this was pilot error

**Comment by santiago** [comment]
@bristoljon how did you solve this issue? i'm experiencing very same behavior where the middleware is not returning back to the saga and just hangs there.
I've got a very simple workflow where UI triggers a saga, this saga does a `call` on **another saga** and then within the latter I make a call to an async function. The async function is called and executed successfully but the middleware just does not come back.
Is it wrong practice to have a saga that `call`s another saga that in turn performs another `call` to a function returning a Promise, like say axios.post?

**Comment by rhysbrettbowen** [comment]
@bristoljon I found a case where `yield*` works but `yield call` won't proceed. It happens when calling a generator that has a fork of a generator with a `takeLatest` in it: https://codesandbox.io/s/basic-redux-saga-n4qb4?fontsize=14
